Client/Frontend/Settings/CustomSearchViewController.swift

```diff
@@ -0,0 +1,206 @@
+                return deferred
+        }
+        FaviconFetcher.fetchFavImageForURL(forURL: url, profile: profile).uponQueue(dispatch_get_main_queue()) { result in
+            let image = result.successValue ?? FaviconFetcher.getDefaultFavicon(url)
+            let engine = OpenSearchEngine(engineID: nil, shortName: name, image: image, searchTemplate: template, suggestTemplate: nil, isCustomEngine: true)
+
+            //Make sure a valid scheme is used
+            let url = engine.searchURLForQuery("test")
+            let maybe = (url == nil) ? Maybe(failure: FaviconError()) : Maybe(success: engine)
+            deferred.fill(maybe)
+        }
+        return deferred
+    }
+
+    func getSearchTemplate(withString query: String) -> String? {
+        let SearchTermComponent = "%s"      //Placeholder in User Entered String
+        let placeholder = "{searchTerms}"   //Placeholder looked for when using Custom Search Engine in OpenSearch.swift
+
+        if query.contains(SearchTermComponent) {
+            return query.stringByReplacingOccurrencesOfString(SearchTermComponent, withString: placeholder)
+        }
+        return nil
+    }
+
+    override func generateSettings() -> [SettingSection] {
+
+        func URLFromString(string: String?) -> NSURL? {
+            guard let string = string else {
+                return nil
+            }
+            return NSURL(string: string)
+        }
+        
+        let titleField = CustomSearchEngineField(placeholder: "Title", settingDidChange: {fieldText in
+            self.engineTitle = fieldText!
+            }, settingIsValid: { text in
+                if text != nil && text != "" {
+                    return true
+                } else {
+                    return false
+                }
+        })
+        titleField.textField.accessibilityIdentifier = "customEngineTitle"
+        
+        let urlField = CustomSearchEngineField(placeholder: "URL (Replace Query with %s)", settingDidChange: {fieldText in
+            self.urlString = fieldText
+            }, settingIsValid: { text in
+                //Can check url text text validity here.
+                return true
+        })
+        urlField.textField.accessibilityIdentifier = "customEngineUrl"
+        
+        let basicSettings: [Setting] = [ titleField, urlField]
+        
+        let settings: [SettingSection] = [
+            SettingSection(title: NSAttributedString(string: "Custom Engine URL"), children: basicSettings),
+            SettingSection(children: [
+            ButtonSetting(title: NSAttributedString(string: "Save"), accessibilityIdentifier: "saveCustomEngine", onClick: addCustomSearchEngine())
+                ])
+        ]
+        
+        return settings
+    }
+    
+    func addCustomSearchEngine() -> (UINavigationController? -> ()) {
+        return { nav in
+            self.view.endEditing(true)
+            self.spinnerView.startAnimating()
```

Comment:
`startAnimating` is also called in `addSearchEngine`. We probably don't need to call it twice.

I would also try to avoid the `if` below since we should always know the string. Just to avoid the potential case of the optional check failing and being in an infinite spinner state.